gp_align/analysis.py

```diff
@@ -0,0 +1,212 @@
+    LOGGER.info("%d images in the series.", len(images))
+    pool = multiprocessing.Pool(processes=num_proc)
+    LOGGER.debug("Submitting tasks...")
+    result_iter = pool.imap_unordered(analyze_image, images)
+    pool.close()
+    with tqdm(total=len(images)) as pbar:
+        for res in result_iter:
+            if "error" in res:
+                LOGGER.error("Image '%s' produced the following error: %s.",
+                             res["filename"], res["error"])
+            else:
+                for plate, row in iteritems(res):
+                    data.setdefault(plate, list()).append(row)
+            pbar.update()
+    pool.join()
+
+    for plate, plate_data in iteritems(data):
+        LOGGER.debug("Plate '%s' has %d rows and %d columns.",
+                     plate, len(plate_data), len(plate_data[0]))
+
+    output = dict()
+    columns = analyze_image.config["well_names"]
+    columns.insert(0, "time")
+    for plate, plate_data in iteritems(data):
+        plate_df = DataFrame(plate_data)
+        assert len(plate_df.columns) == len(columns), "{:d} != {:d}".format(
+            len(plate_df.columns), len(columns))
+        plate_df.sort_values("time", inplace=True)
+        plate_df["time"] -= plate_df["time"].iat[0]
+        plate_df["time"] /= unit
+        output[plate] = plate_df.loc[:, columns]  # order columns
+    return output
+
+
+def configure_run(scanner, plate_type, orientation):
+    config = dict()
+    config["plate_names"] = PLATES[scanner]
+
+    with io.open(join(DATA_DIR, "plate_specs.json"), encoding=None) as file_h:
+        plate_specs = json.load(file_h)
+    rows, columns = plate_specs["rows_and_columns"][str(plate_type)]
+    config["rows"] = rows
+    config["columns"] = columns
+    LOGGER.debug("Plate type %d has %d rows and %d columns.", plate_type,
+                 rows, columns)
+
+    calibration_name_left = "calibration_type_{:d}_left".format(plate_type)
+    config["left_image"] = calibrate_image(join(
+        DATA_DIR, calibration_name_left + ".png"))
+
+    calibration_name_right = "calibration_type_{:d}_right".format(plate_type)
+    config["right_image"] = calibrate_image(join(
+        DATA_DIR, calibration_name_right + ".png"))
+
+    config["well_names"] = well_names(rows, columns, orientation)
+    config["plate_size"] = plate_specs["plate_size"]
+    config["left_positions"] = plate_specs[
+        "plate_positions"][calibration_name_left]
+    config["right_positions"] = plate_specs[
+        "plate_positions"][calibration_name_right]
+
+    analyze_image.config = config
+
+
+def calibrate_image(filename):
+    """Return a normalized gray scale image."""
+    LOGGER.debug(filename)
+    image = rgb2grey(imread(filename))  # rgb2gray can be a noop
+    image = image / image.max()
+    return canny(image, sigma=CANNY_SIGMA)
+
+
+def analyze_image(filename):
+    """Analyze all wells from all trays in one image."""
+    LOGGER.debug(filename)
+    config = analyze_image.config
```

Comment:
This seems convoluted and slightly hacky. Is this a normal design pattern? Maybe it would be more concise to just pass the config dict as an argument to `analyze_image`

Reply:
It's not great. I agree. The point is that this function is basically the multiprocessing worker and the config always contains the same information. I'll take a look whether it has a huge impact to pass it in as an argument.